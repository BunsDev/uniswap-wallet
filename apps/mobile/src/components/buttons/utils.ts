import { withSequence, withSpring, WithSpringConfig } from 'react-native-reanimated'
import { ButtonEmphasis, ButtonSize } from 'src/components/buttons/Button'
import { Theme } from 'ui/src/theme/restyle'

function getButtonColor(emphasis: ButtonEmphasis): keyof Theme['colors'] {
  switch (emphasis) {
    case ButtonEmphasis.Primary:
      return 'accent1'
    case ButtonEmphasis.Secondary:
      return 'surface2'
    case ButtonEmphasis.Tertiary:
      return 'none'
    case ButtonEmphasis.Detrimental:
      return 'DEP_accentCriticalSoft'
    case ButtonEmphasis.Warning:
      return 'DEP_accentWarningSoft'
  }
}

function getButtonTextColor(emphasis: ButtonEmphasis): keyof Theme['colors'] {
  switch (emphasis) {
    case ButtonEmphasis.Primary:
      return 'sporeWhite'
    case ButtonEmphasis.Secondary:
    case ButtonEmphasis.Tertiary:
      return 'neutral1'
    case ButtonEmphasis.Detrimental:
      return 'statusCritical'
    case ButtonEmphasis.Warning:
      return 'DEP_accentWarning'
  }
}

function getButtonBorderColor(emphasis: ButtonEmphasis): keyof Theme['colors'] {
  switch (emphasis) {
    case ButtonEmphasis.Tertiary:
      return 'surface3'
    default:
      return 'none'
  }
}

function getButtonTextSizeVariant(size: ButtonSize): keyof Theme['textVariants'] {
  switch (size) {
    case ButtonSize.Large:
      return 'buttonLabelLarge'
    case ButtonSize.Medium:
      return 'buttonLabelMedium'
    case ButtonSize.Small:
      return 'buttonLabelSmall'
  }
}

function getButtonPaddingY(size: ButtonSize): keyof Theme['spacing'] {
  switch (size) {
    case ButtonSize.Large:
      return 'spacing16'
    case ButtonSize.Medium:
      return 'spacing12'
    case ButtonSize.Small:
      return 'spacing8'
  }
}

function getButtonPaddingX(size: ButtonSize): keyof Theme['spacing'] {
  switch (size) {
    case ButtonSize.Large:
      return 'spacing16'
    case ButtonSize.Medium:
      return 'spacing12'
    case ButtonSize.Small:
      return 'spacing8'
  }
}

function getButtonIconPadding(size: ButtonSize): keyof Theme['spacing'] {
  switch (size) {
    case ButtonSize.Large:
      return 'spacing12'
    case ButtonSize.Medium:
      return 'spacing8'
    case ButtonSize.Small:
      return 'spacing4'
  }
}

function getButtonBorderRadius(size: ButtonSize): keyof Theme['borderRadii'] {
  switch (size) {
    case ButtonSize.Large:
      return 'rounded20'
    case ButtonSize.Medium:
      return 'rounded16'
    case ButtonSize.Small:
      return 'rounded8'
  }
}

function getButtonIconSize(size: ButtonSize): keyof Theme['iconSizes'] {
  switch (size) {
    case ButtonSize.Large:
      return 'icon24'
    case ButtonSize.Medium:
      return 'icon20'
    case ButtonSize.Small:
      return 'icon16'
  }
}

export function getButtonProperties(
  emphasis: ButtonEmphasis,
  size: ButtonSize
): {
  backgroundColor: keyof Theme['colors']
  textColor: keyof Theme['colors']
  textVariant: keyof Theme['textVariants']
  paddingX: keyof Theme['spacing']
  paddingY: keyof Theme['spacing']
  borderRadius: keyof Theme['borderRadii']
  borderColor: keyof Theme['colors']
  iconPadding: keyof Theme['spacing']
  iconSize: keyof Theme['iconSizes']
} {
  const backgroundColor = getButtonColor(emphasis)
  const textColor = getButtonTextColor(emphasis)
  const textVariant = getButtonTextSizeVariant(size)
  const paddingX = getButtonPaddingX(size)
  const paddingY = getButtonPaddingY(size)
  const borderColor = getButtonBorderColor(emphasis)
  const borderRadius = getButtonBorderRadius(size)
  const iconPadding = getButtonIconPadding(size)
  const iconSize = getButtonIconSize(size)

  return {
    backgroundColor,
    textColor,
    textVariant,
    paddingX,
    paddingY,
    borderRadius,
    borderColor,
    iconPadding,
    iconSize,
  }
}

export function pulseAnimation(
  activeScale: number,
  spingAnimationConfig: WithSpringConfig = { damping: 1, stiffness: 200 }
): number {
  'worklet'
  return withSequence(
    withSpring(activeScale, spingAnimationConfig),
    withSpring(1, spingAnimationConfig)
  )
}
