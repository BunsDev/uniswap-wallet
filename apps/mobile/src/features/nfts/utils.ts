export const getNFTAssetKey = (address: Address, token_id: string): string => {
  // Backend returns both checksummed and non-checksummed addresses
  // so we need to lowercase it to be able to compare them
  return `nftItem.${address.toLowerCase()}.${token_id}`
}
