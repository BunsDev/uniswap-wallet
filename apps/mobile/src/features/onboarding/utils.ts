export enum ImportType {
  CreateNew = 'CreateNew',
  CreateAdditional = 'CreateAdditional',
  NotYetSelected = 'NotYetSelected',
  SeedPhrase = 'SeedPhrase',
  Watch = 'Watch',
  Restore = 'Restore',
}

export enum OnboardingEntryPoint {
  Sidebar = 'Sidebar',
  FreshInstallOrReplace = 'FreshInstallOrReplace',
}
