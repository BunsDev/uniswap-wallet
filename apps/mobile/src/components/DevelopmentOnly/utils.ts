import { useEffect, useState } from 'react'
import { useAppDispatch } from 'src/app/hooks'
import { pushNotification } from 'wallet/src/features/notifications/slice'
import { useActiveAccountAddressWithThrow } from 'wallet/src/features/wallet/hooks'

export const fakeNotification = {
  type: 2,
  // address: '0x...',
  dappName: 'Uniswap Interface',
  event: 1,
  imageUrl: 'https://app.uniswap.org/favicon.png',
  hideDelay: 3000,
}

// easiest to use inside NotificationToastWrapper before any returns
export const useFakeNotification = (ms?: number): void => {
  const [sent, setSent] = useState(false)
  const dispatch = useAppDispatch()
  const activeAddress = useActiveAccountAddressWithThrow()

  useEffect(() => {
    setSent(false)
  }, [ms])

  useEffect(() => {
    if (!sent && activeAddress) {
      dispatch(
        pushNotification({
          ...fakeNotification,
          hideDelay: ms ?? fakeNotification.hideDelay,
          address: activeAddress,
        })
      )
      setSent(true)
    }
  }, [activeAddress, dispatch, ms, sent])
}
