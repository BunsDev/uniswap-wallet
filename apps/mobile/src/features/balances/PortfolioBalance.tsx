import React, { useEffect, useState } from 'react'
import AnimatedNumber from 'src/components/AnimatedNumber'
import { DeprecatedMobileOnlyFlex } from 'src/components/layout'
import { iconSizes } from 'ui/src/theme'
import { formatUSDPrice, NumberType } from 'utilities/src/format/format'
import { RelativeChange } from 'wallet/src/components/text/RelativeChange'
import { PollingInterval } from 'wallet/src/constants/misc'
import { isWarmLoadingStatus } from 'wallet/src/data/utils'
import { usePortfolioBalancesQuery } from 'wallet/src/data/__generated__/types-and-hooks'

interface PortfolioBalanceProps {
  owner: Address
}

export function PortfolioBalance({ owner }: PortfolioBalanceProps): JSX.Element {
  const { data, loading, networkStatus } = usePortfolioBalancesQuery({
    variables: { ownerAddress: owner },
    // TransactionHistoryUpdater will refetch this query on new transaction.
    // No need to be super aggressive with polling here.
    pollInterval: PollingInterval.Normal,
    notifyOnNetworkStatusChange: true,
    // This is better than using network status to check, because doing it that way we would have to wait
    // for the network status to go back to "ready", which results in the numbers updating, and _then_ the
    // shimmer disappearing. Using onCompleted it disappears at the same time as the data loads in.
    onCompleted: () => {
      setIsWarmLoading(false)
    },
  })

  const [isWarmLoading, setIsWarmLoading] = useState(false)
  const isLoading = loading && !data

  useEffect(() => {
    if (!!data && isWarmLoadingStatus(networkStatus)) {
      setIsWarmLoading(true)
    }
  }, [data, networkStatus])

  const portfolioBalance = data?.portfolios?.[0]
  const portfolioChange = portfolioBalance?.tokensTotalDenominatedValueChange
  const totalBalance = portfolioBalance?.tokensTotalDenominatedValue?.value

  return (
    <DeprecatedMobileOnlyFlex gap="spacing4">
      <AnimatedNumber
        colorIndicationDuration={2000}
        loading={isWarmLoading || isLoading}
        loadingPlaceholderText="$00000.00"
        value={formatUSDPrice(totalBalance, NumberType.PortfolioBalance)}
      />
      <RelativeChange
        absoluteChange={portfolioChange?.absolute?.value}
        arrowSize={iconSizes.icon20}
        change={portfolioChange?.percentage?.value}
        loading={isWarmLoading || isLoading}
        negativeChangeColor={isWarmLoading ? '$neutral2' : '$statusCritical'}
        positiveChangeColor={isWarmLoading ? '$neutral2' : '$statusSuccess'}
        variant="body1"
      />
    </DeprecatedMobileOnlyFlex>
  )
}
