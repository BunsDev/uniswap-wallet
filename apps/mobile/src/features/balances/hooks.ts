import { useCallback, useMemo } from 'react'
import { useTranslation } from 'react-i18next'
import { NativeSyntheticEvent } from 'react-native'
import { ContextMenuAction, ContextMenuOnPressNativeEvent } from 'react-native-context-menu-view'
import { useAppDispatch, useAppSelector } from 'src/app/hooks'
import { selectTokensVisibility } from 'src/features/favorites/selectors'
import { toggleTokenVisibility, TokenVisibility } from 'src/features/favorites/slice'
import { useSelectLocalTxCurrencyIds } from 'src/features/transactions/hooks'
import { EMPTY_ARRAY } from 'wallet/src/constants/misc'
import { PortfolioBalance } from 'wallet/src/features/dataApi/types'
import { pushNotification } from 'wallet/src/features/notifications/slice'
import { AppNotificationType } from 'wallet/src/features/notifications/types'
import {
  useAccounts,
  useSelectAccountHideSmallBalances,
  useSelectAccountHideSpamTokens,
} from 'wallet/src/features/wallet/hooks'
import { CurrencyId } from 'wallet/src/utils/currencyId'
import { ONE_SECOND_MS } from 'wallet/src/utils/time'

interface TokenMenuParams {
  currencyId: CurrencyId
  owner: Address
  isSpam: Maybe<boolean>
  isNative: Maybe<boolean>
  balanceUSD: Maybe<number>
  tokenSymbolForNotification?: Nullable<string>
}

const HIDE_SMALL_USD_BALANCES_THRESHOLD = 1

export const HIDDEN_TOKEN_BALANCES_ROW = 'HIDDEN_TOKEN_BALANCES_ROW'

/**
 * Checks if a token balance should be hidden.
 *
 * @param hideSpamTokens - Indicates whether to hide spam tokens.
 * @param hideSmallBalances - Indicates whether to hide tokens with small balances.
 * @param isSpam - The spam status of the token, which could be undefined.
 * @param balanceUSD - The balance of the token in USD, which could be undefined.
 * @param tokenVisibility - Optional parameter that includes data about token's visibility.
 * @param isSentOrSwappedLocally - Indicates if the token has been sent or swapped locally.
 *
 * @returns {boolean} - Returns true if the token balance should be hidden, false otherwise.
 *
 * @example
 * const isHidden = isTokenBalanceHidden({ hideSpamTokens, hideSmallBalances, isSpam, balanceUSD });
 */
function isTokenBalanceHidden({
  hideSpamTokens,
  hideSmallBalances,
  isSpam,
  isNative,
  balanceUSD,
  tokenVisibility,
  isSentOrSwappedLocally,
}: {
  hideSpamTokens: boolean
  hideSmallBalances: boolean
  isSpam: Maybe<boolean>
  isNative: Maybe<boolean>
  balanceUSD: Maybe<number>
  tokenVisibility?: TokenVisibility
  isSentOrSwappedLocally: Maybe<boolean>
}): boolean {
  // If user has explicity set visibility then always follow their preference
  if (tokenVisibility !== undefined) return !tokenVisibility.isVisible

  // If the user has sent or swapped a given token before then do not hide it by default
  if (isSentOrSwappedLocally) return false

  const shouldHideSpam = hideSpamTokens && isSpam
  const isSmallBalance = !balanceUSD || balanceUSD < HIDE_SMALL_USD_BALANCES_THRESHOLD
  const shouldHideSmallBalance = hideSmallBalances && isSmallBalance && isNative === false

  // Hide a token by default if it is spam or a non-native token with a small balance
  return shouldHideSpam || shouldHideSmallBalance
}

function useAccountTokensVisibilitySettings(owner: Address): {
  hideSpamTokens: boolean
  hideSmallBalances: boolean
  accountTokensVisibility?: Record<string, TokenVisibility>
  sentOrSwappedLocally: Record<string, boolean>
} {
  const hideSmallBalances = useSelectAccountHideSmallBalances(owner)
  const hideSpamTokens = useSelectAccountHideSpamTokens(owner)
  const sentOrSwappedLocally = useSelectLocalTxCurrencyIds(owner)
  const tokensVisibility = useAppSelector(selectTokensVisibility)
  return {
    hideSmallBalances,
    hideSpamTokens,
    accountTokensVisibility: tokensVisibility[owner],
    sentOrSwappedLocally,
  }
}

// Provide context menu related data for token balance
export function useTokenBalanceContextMenu({
  currencyId,
  owner,
  isSpam,
  isNative,
  balanceUSD,
  tokenSymbolForNotification,
}: TokenMenuParams): {
  menuActions: Array<ContextMenuAction & { onPress: () => void }>
  onContextMenuPress: (e: NativeSyntheticEvent<ContextMenuOnPressNativeEvent>) => void
} {
  const { t } = useTranslation()
  const dispatch = useAppDispatch()
  const { hideSmallBalances, hideSpamTokens, accountTokensVisibility, sentOrSwappedLocally } =
    useAccountTokensVisibilitySettings(owner)
  const isHidden = isTokenBalanceHidden({
    tokenVisibility: accountTokensVisibility?.[currencyId],
    isSentOrSwappedLocally: sentOrSwappedLocally[currencyId],
    hideSpamTokens,
    hideSmallBalances,
    isSpam,
    isNative,
    balanceUSD,
  })

  const accounts = useAccounts()
  const isLocalAccount = !!accounts[owner]

  const menuActions = useMemo(
    () => [
      ...(isLocalAccount
        ? [
            {
              title: isHidden ? t('Unhide Token') : t('Hide Token'),
              systemIcon: isHidden ? 'eye' : 'eye.slash',
              destructive: !isHidden,
              onPress: (): void => {
                dispatch(
                  toggleTokenVisibility({
                    owner,
                    currencyId: currencyId.toLowerCase(),
                    currentlyVisible: !isHidden,
                  })
                )
                if (tokenSymbolForNotification) {
                  dispatch(
                    pushNotification({
                      type: AppNotificationType.AssetVisibility,
                      visible: !isHidden,
                      hideDelay: 2 * ONE_SECOND_MS,
                      assetName: tokenSymbolForNotification,
                    })
                  )
                }
              },
            },
          ]
        : []),
    ],
    [isLocalAccount, isHidden, t, dispatch, owner, currencyId, tokenSymbolForNotification]
  )

  const onContextMenuPress = useCallback(
    (e: NativeSyntheticEvent<ContextMenuOnPressNativeEvent>): void => {
      menuActions[e.nativeEvent.index]?.onPress?.()
    },
    [menuActions]
  )

  return { menuActions, onContextMenuPress }
}

/**
 * Custom hook to group Token Balances fetched from API to visible and hidden.
 *
 * @param balancesById - An object where keys are token ids and values are the corresponding balances. May be undefined.
 * @param expandHiddenTokens - Boolean flag to indicate if hidden tokens should be expanded.
 * @param owner - The owner address for which token balances are managed.
 *
 * @returns {object} An object containing two fields:
 *  - `tokens`: an array of tokens which could be a mix of PortfolioBalance instances or string (presumably token ids).
 *  - `numHidden`: the number of hidden tokens.
 *
 * @example
 * const { tokens, numHidden } = useGroupTokenBalancesByVisibility({ balancesById, expandHiddenTokens, owner });
 */
export function useTokenBalancesGroupedByVisibility({
  balancesById,
  expandHiddenTokens,
  owner,
}: {
  balancesById?: Record<string, PortfolioBalance>
  expandHiddenTokens: boolean
  owner: Address
}): {
  tokens: Array<PortfolioBalance | string>
  numHidden: number
} {
  const { hideSmallBalances, hideSpamTokens, accountTokensVisibility, sentOrSwappedLocally } =
    useAccountTokensVisibilitySettings(owner)

  return useMemo(() => {
    if (!balancesById) {
      return { tokens: EMPTY_ARRAY, numHidden: 0 }
    }

    const { shown, hidden } = Object.values(balancesById).reduce<{
      shown: PortfolioBalance[]
      hidden: PortfolioBalance[]
    }>(
      (acc, balance) => {
        const isHidden = isTokenBalanceHidden({
          tokenVisibility: accountTokensVisibility?.[balance.currencyInfo.currencyId],
          isSentOrSwappedLocally: sentOrSwappedLocally[balance.currencyInfo.currencyId],
          hideSpamTokens,
          hideSmallBalances,
          balanceUSD: balance.balanceUSD,
          isSpam: balance.currencyInfo.isSpam,
          isNative: balance.currencyInfo.currency.isNative,
        })

        if (isHidden) {
          acc.hidden.push(balance)
        } else {
          acc.shown.push(balance)
        }
        return acc
      },
      { shown: [], hidden: [] }
    )
    return {
      tokens: [
        ...shown,
        ...(hidden.length ? [HIDDEN_TOKEN_BALANCES_ROW] : []),
        ...(expandHiddenTokens ? hidden : []),
      ],
      numHidden: hidden.length,
    }
  }, [
    balancesById,
    expandHiddenTokens,
    accountTokensVisibility,
    sentOrSwappedLocally,
    hideSpamTokens,
    hideSmallBalances,
  ])
}
