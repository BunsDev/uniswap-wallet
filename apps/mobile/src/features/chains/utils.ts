import { useMemo } from 'react'
import { SelectEffect } from 'redux-saga/effects'
import { appSelect, useAppSelector } from 'src/app/hooks'
import {
  ALL_SUPPORTED_CHAIN_IDS,
  ChainId,
  ChainIdTo,
  ChainState,
} from 'wallet/src/constants/chains'

export function useActiveChainIds(): ChainId[] {
  const chains = useAppSelector((state) => state.chains.byChainId)
  return useMemo(() => getSortedActiveChainIds(chains), [chains])
}

export function* selectActiveChainIds(): Generator<SelectEffect, ChainId[], unknown> {
  const chains = yield* appSelect((s) => s.chains.byChainId)
  return getSortedActiveChainIds(chains)
}

// ALL_SUPPORTED_CHAINS is manually sorted by chain TVL
export function getSortedActiveChainIds(chains: ChainIdTo<ChainState>): ChainId[] {
  return ALL_SUPPORTED_CHAIN_IDS.filter((n: ChainId) => !!chains[n]?.isActive)
}
