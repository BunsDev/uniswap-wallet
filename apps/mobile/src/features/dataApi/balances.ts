import { useMemo } from 'react'
import { useActiveAccountAddressWithThrow } from 'src/features/wallet/hooks'
import { usePortfolioBalances } from 'wallet/src/features/dataApi/balances'
import { PortfolioBalance } from 'wallet/src/features/dataApi/types'
import { CurrencyId } from 'wallet/src/utils/currencyId'

/** Helper hook to retrieve balance for a single currency for the active account. */
export function useSingleBalance(_currencyId: CurrencyId): PortfolioBalance | null {
  const address = useActiveAccountAddressWithThrow()
  const { data: portfolioBalances } = usePortfolioBalances(address, /*shouldPoll=*/ false)

  return useMemo(() => {
    if (!portfolioBalances) return null
    return portfolioBalances[_currencyId] ?? null
  }, [portfolioBalances, _currencyId])
}

/** Helper hook to retrieve balances for a set of currencies for the active account. */
export function useMultipleBalances(
  currencies: CurrencyId[] | undefined
): PortfolioBalance[] | null {
  const address = useActiveAccountAddressWithThrow()
  const { data: balances } = usePortfolioBalances(address, /*shouldPoll=*/ false)

  return useMemo(() => {
    if (!currencies || !currencies.length || !balances) return null

    return currencies
      .map((id: CurrencyId) => balances[id] ?? null)
      .filter((x): x is PortfolioBalance => Boolean(x))
  }, [balances, currencies])
}
