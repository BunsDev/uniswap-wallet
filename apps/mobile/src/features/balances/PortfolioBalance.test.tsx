import { MockedResponse } from '@apollo/client/testing'
import React from 'react'
import { PortfolioBalance } from 'src/features/balances/PortfolioBalance'
import { Portfolios } from 'src/test/gqlFixtures'
import { render, screen } from 'src/test/test-utils'
import {
  PortfolioBalancesDocument,
  PortfolioBalancesQuery,
} from 'wallet/src/data/__generated__/types-and-hooks'
import { formatUSDPrice, NumberType } from 'wallet/src/utils/format'

const mock: MockedResponse<PortfolioBalancesQuery> = {
  request: {
    query: PortfolioBalancesDocument,
    variables: {
      ownerAddress: Portfolios[0].ownerAddress,
    },
  },
  result: {
    data: {
      portfolios: Portfolios,
    },
  },
}

describe(PortfolioBalance, () => {
  it('renders without error', async () => {
    const tree = render(<PortfolioBalance owner={Portfolios[0].ownerAddress} />, { mocks: [mock] })

    // loading
    expect(await screen.findByText('$000.00')).toBeDefined()
    expect(tree.toJSON()).toMatchSnapshot()

    // success
    expect(
      await screen.findByText(
        formatUSDPrice(
          Portfolios[0].tokensTotalDenominatedValue?.value,
          NumberType.PortfolioBalance
        )
      )
    ).toBeDefined()
    expect(tree.toJSON()).toMatchSnapshot()
  })
})
