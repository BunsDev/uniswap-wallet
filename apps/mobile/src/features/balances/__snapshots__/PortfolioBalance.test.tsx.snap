// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PortfolioBalance renders without error 1`] = `
<View
  onLayout={[Function]}
>
  <View
    style={
      [
        {
          "alignItems": "stretch",
          "display": "flex",
          "flexDirection": "column",
          "justifyContent": "flex-start",
        },
      ]
    }
  >
    <View
      onLayout={[Function]}
    >
      <View
        style={
          [
            {
              "alignItems": "center",
              "display": "flex",
              "flexDirection": "row",
            },
          ]
        }
      >
        <View
          style={
            [
              {
                "alignItems": "center",
                "display": "flex",
                "flexDirection": "row",
                "position": "relative",
              },
            ]
          }
        >
          <View
            accessibilityElementsHidden={true}
            importantForAccessibility="no-hide-descendants"
          >
            <Text
              adjustsFontSizeToFit={false}
              allowFontScaling={true}
              maxFontSizeMultiplier={1.2}
              numberOfLines={1}
              style={
                [
                  {
                    "color": "#0E1524",
                    "display": "flex",
                    "fontFamily": "Inter-SemiBold",
                    "fontSize": 48,
                    "fontWeight": "600",
                    "lineHeight": 48,
                    "opacity": 0,
                  },
                ]
              }
            >
              $000.00
            </Text>
          </View>
          <View
            style={
              [
                {
                  "backgroundColor": "#E8ECFB",
                  "borderRadius": 4,
                  "bottom": "5%",
                  "display": "flex",
                  "left": 0,
                  "position": "absolute",
                  "right": 0,
                  "top": "5%",
                },
              ]
            }
          />
        </View>
      </View>
    </View>
    <View
      style={
        [
          {
            "display": "flex",
            "flexGrow": 0,
            "flexShrink": 0,
            "height": 4,
          },
        ]
      }
    />
    <View
      style={
        [
          {
            "alignItems": "center",
            "display": "flex",
            "flexDirection": "row",
            "justifyContent": "flex-start",
          },
        ]
      }
    >
      <View
        onLayout={[Function]}
      >
        <View
          style={
            [
              {
                "alignItems": "center",
                "display": "flex",
                "flexDirection": "row",
              },
            ]
          }
        >
          <View
            style={
              [
                {
                  "alignItems": "center",
                  "display": "flex",
                  "flexDirection": "row",
                  "position": "relative",
                },
              ]
            }
          >
            <View
              accessibilityElementsHidden={true}
              importantForAccessibility="no-hide-descendants"
            >
              <Text
                allowFontScaling={true}
                maxFontSizeMultiplier={1.4}
                style={
                  [
                    {
                      "color": "#5D6785",
                      "display": "flex",
                      "fontFamily": "Inter-Medium",
                      "fontSize": 17,
                      "lineHeight": 24,
                      "opacity": 0,
                    },
                  ]
                }
              >
                $0.00 (0.00)%
              </Text>
            </View>
            <View
              style={
                [
                  {
                    "backgroundColor": "#E8ECFB",
                    "borderRadius": 4,
                    "bottom": "5%",
                    "display": "flex",
                    "left": 0,
                    "position": "absolute",
                    "right": 0,
                    "top": "5%",
                  },
                ]
              }
            />
          </View>
        </View>
      </View>
    </View>
  </View>
</View>
`;

exports[`PortfolioBalance renders without error 2`] = `
<View
  onLayout={[Function]}
>
  <View
    style={
      [
        {
          "alignItems": "stretch",
          "display": "flex",
          "flexDirection": "column",
          "justifyContent": "flex-start",
        },
      ]
    }
  >
    <Text
      adjustsFontSizeToFit={true}
      allowFontScaling={true}
      maxFontSizeMultiplier={1.2}
      numberOfLines={1}
      style={
        [
          {
            "color": "#0E1524",
            "display": "flex",
            "fontFamily": "Inter-SemiBold",
            "fontSize": 48,
            "fontWeight": "600",
            "lineHeight": 48,
          },
        ]
      }
    >
      $55
      <Text
        allowFontScaling={true}
        maxFontSizeMultiplier={1.2}
        style={
          [
            {
              "color": "#888FAB",
              "display": "flex",
              "fontFamily": "Inter-SemiBold",
              "fontSize": 40,
              "lineHeight": 48,
            },
          ]
        }
      >
        .
        00
      </Text>
    </Text>
    <View
      style={
        [
          {
            "display": "flex",
            "flexGrow": 0,
            "flexShrink": 0,
            "height": 4,
          },
        ]
      }
    />
    <View
      style={
        [
          {
            "alignItems": "center",
            "display": "flex",
            "flexDirection": "row",
            "justifyContent": "flex-start",
          },
        ]
      }
    >
      <svg
        color="#40B66B"
        data-file-name="SvgArrowChange"
        height={20}
        strokeWidth={2}
        style={
          {
            "transform": [
              {
                "rotate": "180deg",
              },
            ],
          }
        }
        width={20}
      />
      <View
        style={
          [
            {
              "display": "flex",
              "flexGrow": 0,
              "flexShrink": 0,
              "height": 2,
              "width": 2,
            },
          ]
        }
      />
      <Text
        allowFontScaling={true}
        maxFontSizeMultiplier={1.4}
        style={
          [
            {
              "color": "#5D6785",
              "display": "flex",
              "fontFamily": "Inter-Medium",
              "fontSize": 17,
              "lineHeight": 24,
            },
          ]
        }
      >
        $5.00 (0.05%)
      </Text>
    </View>
  </View>
</View>
`;
